File: ch5/python/code5_2.py
def edit_distance(str1, str2, m, n):
    edit_distances = [[0] * (n + 1) for i in range(0, m + 1)]

    for j in range(0, n + 1):   # 제일 위 행
        edit_distances[0][j] = j
    
    for i in range(0, m + 1):   # 제일 왼쪽 열
        edit_distances[i][0] = i

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            # 두 글자가 같다면
            if str1[i - 1] == str2[j - 1]:
                edit_distances[i][j] = edit_distances[i - 1][j - 1]
            # 두 글자가 다르다면
            else:
                edit_distances[i][j] = min(edit_distances[i][j - 1], \
                                            edit_distances[i - 1][j], \
                                            edit_distances[i - 1][j - 1]) + 1

    # 완성된 EditD 행렬을 출력해봅니다.
    for i in range(0, m + 1):
        for j in range(0, n + 1):
            print('%2d' % edit_distances[i][j], end = '')
        print()
    
    return edit_distances[m][n]

str1, str2 = input('두 단어를 입력하세요 : ').split()
print('두 단어 [%s] [%s] 사이의 최소 교정 비용은 %d입니다.' % \
    (str1, str2, edit_distance(str1, str2, len(str1), len(str2))))

File: ch5/c/code5_2.c
#include <stdio.h>
#include <string.h>

int getMinimum(int a, int b, int c)
{
  if(a >= c && b >= c)
    return c;
  return a > b? b: a;
}

//#define M 7
//#define N 9
int editDistance(char* str1, char* str2, int m, int n)
{
  int EditD[m + 1][n + 1];

  for(int j = 0; j <= n; j++) // 제일 위 행
    EditD[0][j] = j;

  for(int i = 0; i <= m; i++) // 제일 왼쪽 열
    EditD[i][0] = i;

  for(int i = 1; i <= m; i++)
  {
    for(int j = 1; j <= n; j++)
    {
      // 두 글자가 같다면
      if(str1[i - 1] == str2[j - 1])
        EditD[i][j] = EditD[i - 1][j - 1];
      // 두 글자가 다르다면
      else
        EditD[i][j] = getMinimum(EditD[i][j - 1],
                                 EditD[i - 1][j],
                                 EditD[i - 1][j - 1]) + 1;
      
    }
  }

  // EditD 행렬 출력
  for(int i = 0; i <= m; i++)
  {
    for(int j = 0; j <= n; j++)
    {
      printf("%2d", EditD[i][j]);
    }
    printf("\n");
  }
  return EditD[m][n];
}

int main()
{
  char str1[32], str2[32];
  printf("두 단어를 입력하세요 : ");
  scanf("%s %s", str1, str2);
  printf("두 단어 [%s] [%s] 사이의 최소 교정 비용은 %d입니다.\n", 
    str1, str2, editDistance(str1, str2, strlen(str1), strlen(str2)));
  return 0;    
}
